wxBitmap multi-resolution API testbed

csomor opened the pull request:
This is a testbed to find out the best API for supporting multiple representations of the same content, for best support of high-dpi environments

csomor committed:
multi-res API

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -134,6 +134,66 @@ private :
 };
 
+typedef wxObjectDataPtr<wxBitmapRepresentationData> wxBitmapRepresentationDataPtr;
+WX_DEFINE_TYPEARRAY( wxObjectDataPtr<wxBitmapRepresentationData>, wxBitmapRepresentationArray, dummy );
```
We should just use `std::vector<>` here.

csomor replied:
Thanks, I wanted to have an implementation which could be copied to all ports, so std::vector is element of our 'baseline', STL always on ?

vadz replied:
To be maximally safe let's use `wxVector<>` if it's to be used in all ports. TIA!

csomor replied:
ok, thanks, I'll do so

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -134,6 +134,66 @@ private :
+    void AddRepresentations(const wxBitmap& other)
+    {
+        const wxBitmapRefData* otherData = other.GetBitmapData() ;
+        for ( size_t i = 0 ; i < otherData->m_representations.size() ; ++i )
+        {
+            m_representations.push_back(otherData->m_representations.at(i));
+        }
```
This should be just `m_reps.insert(m_reps.end(), otherReps.begin(), otherReps.end())`

csomor replied:
but this is not defined with wxVector, isn't it ?

vadz replied:
Yes, an oversight :-( I think we should add it to `wxVector<>` too, it's simple enough...

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -134,6 +134,66 @@ private :
+
+    wxBitmapRepresentationData* GetDefaultRepresentation()
+    {
+        return m_representations.at(0).get();
```
Throwing from here when there are no representations doesn't seem like a good idea (if it's intentional).

We should probably just return `NULL` if the vector is empty (and if it isn't, it's safe to use `[0]` instead of `.at(0)`)

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -134,6 +134,66 @@ private :
+            if ( rep->GetWidth() == dimensions.x && rep->GetHeight() == dimensions.y )
+                return rep;
+        }
+        return m_representations.at(0).get();
```
```suggestion
        return GetDefaultRepresentation();
```

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -134,6 +134,66 @@ private :
+        }
+    }
+
+    wxBitmapRepresentationData* GetDefaultRepresentation()
```
```suggestion
    wxBitmapRepresentationData* GetDefaultRepresentation() const
```

vadz commented on src/osx/core/bitmap.cpp:
```diff
@@ -949,15 +1026,18 @@ wxBitmap wxBitmap::GetSubBitmap(const wxRect &rect) const
     int destwidth = rect.width*scale ;
     int destheight = rect.height*scale ;
 
+    const wxBitmapRepresentationData* srcrep = GetBitmapData()->GetDefaultRepresentation();
+    wxBitmapRepresentationData* destrep = ret.GetBitmapData()->GetDefaultRepresentation();
```
It's probably not feasible to apply this to all representations, but maybe worth adding a comment at least saying this.

vadz reviewed (commented):
No real comments, but then I think we should decide what to do with `CreateScaled()` and `GetScaledXXX()` first, as the changes proposed in my ML message will probably affect this PR too.

csomor committed:
switching to wxVector, applying suggested changes

csomor commented:
Now we could define which wxBitmap methods work only if there is only one representation. I'd add an assert for size == 1, think this is true for most of them ...

Another point is whether we need some kind of check that all representations have the same aspect ratio.

Scale being doubles, I thought that maybe DPI would be a better choice for certain parameters, as these would be int.

vadz commented:
> Now we could define which wxBitmap methods work only if there is only one representation. I'd add an assert for size == 1, think this is true for most of them ...

Ideally I'd rather add optional parameter allowing to select the representation to operate on.

> Another point is whether we need some kind of check that all representations have the same aspect ratio.

I don't think so. What would we do if it fails, anyhow?
 
> Scale being doubles, I thought that maybe DPI would be a better choice for certain parameters, as these would be int.

I'd go with `wxBitmapRepr` class that could potentially be constructed from different things (including nothing, i.e. default-constructed object of this class would select the default representation).

csomor committed:
first take at multi-res API

introducing bitmap scale and bitmap representation class
changing type of bitmap.cpp to bitmap.mm

csomor committed:
fixing compilation on iOS

csomor committed:
reducing API surface, moving obj-c methods to bitmap.mm

csomor committed:
fixing non-precomp and multi-lib build

csomor committed:
missing commit for moved methods

csomor committed:
use new methods in macOS code parts

csomor committed:
method is not overridden …

csomor committed:
adding NSImage / UIImage creation

csomor committed:
fixing iphone build

csomor commented:
@vadz I tried to bring together the topics we discussed while maintaining functionality on macOS ... so we can discuss things in code

As a side-note, when going through the conversion from system delivered NSImage instances, I had to realize, that they have some note of system appearance - ie they have different images for dark or light mode, unfortunately this is not accessible via a public api, though I could parse the object's description to find that data all the same. 
So I wonder whether we would want to have an additional tag-string for representations which we might use for selection

csomor committed:
fixing memory leak

vadz commented:
Stefan, thanks for all the changes and sorry for the delay. Except a few trivial comments above (that I can address myself if you wish), I don't have much to say. I think we should already merge this, to prevent it from going stale, but in order to do it we must, of course, document the new API and fix the (numerous) CI build failures. Should I do this or do you prefer to do it yourself?

csomor commented:
Thanks, Vadim. I'd be glad if you could do it. If I can assist somehow, please let me know.

csomor committed:
supporting alternate naming scheme (_2x instead of @2x etc.)

csomor committed:
copying over toolbar sample from PR2520

https://github.com/wxWidgets/wxWidgets/pull/2520

csomor committed:
add switches for different configurations under same platform

csomor committed:
adding merge of bitmaps, testing multi-res construction from pngs in c files

csomor committed:
going protected with the representations array

csomor committed:
reducing public API surface, extending toolbar sample

testing at a possible implementation for manually switching bitmap based on best representation

csomor committed:
fixing non-multi-res bitmap path

vadz commented:
I've been looking at rebasing this on the current master and unfortunately this is not trivial because 2add072817 both renames the existing `src/osx/core/bitmap.cpp` to `src/osx/core/bitmap.mm` and changes it and it's difficult to reconcile this with the changes that happened to this file in master since then. I'll have to review this carefully to ensure that nothing gets lost, but without fully understanding the changes it's going to be difficult for me...

csomor commented:
> I've been looking at rebasing this on the current master and unfortunately this is not trivial because [2add072](https://github.com/wxWidgets/wxWidgets/commit/2add07281770bc6f6680bd4231276d6fe5e6ee45) both renames the existing `src/osx/core/bitmap.cpp` to `src/osx/core/bitmap.mm` and changes it and it's difficult to reconcile this with the changes that happened to this file in master since then. I'll have to review this carefully to ensure that nothing gets lost, but without fully understanding the changes it's going to be difficult for me...

how can I help ? you can leave it in any state that fits and I'll update it

vadz commented:
@csomor I think this one can be closed now that #2535 is merged, but please let me know if there is still anything worth preserving here (or just close it yourself). TIA!

csomor commented:
Yes, I'll close it, there's still code that might be useful, but since I won't delete my branch, there's no need for the PR.
